nixos/modules/security/acme/default.nix

```
@@ -929,32 +940,23 @@ in {
           '';
         }
         {
-          assertion = data.dnsProvider == null || data.webroot == null;
-          message = ''
-            Options `security.acme.certs.${cert}.dnsProvider` and
-            `security.acme.certs.${cert}.webroot` are mutually exclusive.
-          '';
-        }
-        {
-          assertion = data.webroot == null || data.listenHTTP == null;
-          message = ''
-            Options `security.acme.certs.${cert}.webroot` and
-            `security.acme.certs.${cert}.listenHTTP` are mutually exclusive.
-          '';
-        }
-        {
-          assertion = data.listenHTTP == null || data.dnsProvider == null;
-          message = ''
-            Options `security.acme.certs.${cert}.listenHTTP` and
-            `security.acme.certs.${cert}.dnsProvider` are mutually exclusive.
-          '';
-        }
-        {
-          assertion = data.dnsProvider != null || data.webroot != null || data.listenHTTP != null;
+          assertion = let
+            boolToInt = x: if false then 0 else 1;
+            sum = l: lib.fold (x: y: x + y) 0 l;
+          in
+            sum (map (x: boolToInt (x != null)) [
+              data.dnsProvider
+              data.webroot
+              data.listenHTTP
+              data.s3Bucket
+            ]) != 1;
```

I find the following much easier to read:

```suggestion
          assertion = lib.length (lib.filter (x: x != null) [
            data.dnsProvider
            data.webroot
            data.listenHTTP
            data.s3Bucket
          ]) != 1;
```